### PR TITLE
Setting elastic password from stored hash

### LINKED ISSUE #75704: Set elastic password from stored hash for packaged installations
During packaged installations we will be generating and storing a new setting, named `autoconfiguration.password_hash` in the `elasticsearch.keystore`. This will be the -to be- password of the elastic user and its value in the keystore will be salted and hashed using the default hashing algorithm of elasticsearch. 

At startup, the node should

- Detect that there is an `autoconfiguration.password_hash` setting in the keystore
- Detect that there is no `.security` index
- Create the document for the elastic built-in user in the security index and set its password to the value it read from  `autoconfiguration.password_hash`

### PR DESCRIPTION
**Pull request by BigPandaToo**
In package installations, we will be generating the password of
the elastic user on installation and we will be storing the hash
of it to the elasticsearch.keystore.
This change will
- attempt to create the security index and elastic user document at the
time of the first successful authentication using the
the "promised" password
- if bootstrap.password setting is detected, won’t attempt to use
autoconfiguration.password_hash setting and proceed with the normal
authentication path
- if elastic user document exists, won't attempt to use
autoconfiguration.password_hash setting and proceed with the normal
authentication path
- if unable  to create the Security Index for any reason, returns
HTTP_SERVER_ERROR to indicate that the cluster is probably not ready yet

Resolves: #75704

### CODE REVIEW

**Comment by BigPandaToo** [comment]
@elasticmachine update branch

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/esnative/ReservedRealm.java
Comment (jkakavas) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/esnative/ReservedRealm.java:231]: we can extract this in a method instead of duplicating the code here

**Review by jkakavas — COMMENTED** [review]
Thanks  Lyudmila, I think this does what we've set out to do in the design doc. I think it's ++ to polish and open as a proper PR when you feel ready for it. We might be able to wire things differently in the ReservedRealm class but I don't have a concrete suggestion now, nor specific concerns around your approach. Let's see what @albertzaharovits thinks also

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/esnative/ReservedRealm.java
Comment (albertzaharovits) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/esnative/ReservedRealm.java:231]: ++ But we don't need a new method even, we can move this logic in the caller:
```
@@ -283,6 +284,11 @@ public class NativeUsersStore {
         });
     }

+    void storeAutoConfigElasticUser(ActionListener<User> listener) {
+        createReservedUser(ElasticUser.NAME, autoConfigUserInfo.hash, RefreshPolicy.IMMEDIATE,
+                listener.map(aVoid -> new ElasticUser(true)));
+    }
+
     /**
      * Asynchronous method to put a user. A put user request without a password hash is treated as an update and will fail with a
      * {@link ValidationException} if the user does not exist. If a password hash is provided, then we issue a update request with an
diff --git a/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/esnative/ReservedRealm.java b/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/esnative/ReservedRealm.java
index 2dea4b8b1f2..65382de98ca 100644
--- a/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/esnative/ReservedRealm.java
+++ b/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/esnative/ReservedRealm.java
@@ -102,6 +102,10 @@ public class ReservedRealm extends CachingUsernamePasswordRealm {
                         if (userInfo.hasEmptyPassword()) {
                             result = AuthenticationResult.terminate("failed to authenticate user [" + token.principal() + "]", null);
                         } else if (userInfo.verifyPassword(token.credentials())) {
+                            if (userInfo == autoconfigUserInfo)) {
+                                nativeUsersStore.storeAutoConfigElasticUser(listener.map(user -> AuthenticationResult.success(user)));
+                                return;
+                            }
                             final User user = getUser(token.principal(), userInfo);
                             logDeprecatedUser(user);
                             result = AuthenticationResult.success(user);
@@ -242,7 +246,12 @@ public class ReservedRealm extends CachingUsernamePasswordRealm {

     private ReservedUserInfo getDefaultUserInfo(String username) {
         if (ElasticUser.NAME.equals(username)) {
-            return bootstrapUserInfo.deepClone();
+            if (autoconfigHashExists && false == bootstrapPasswordExists) {
+                assert autoconfigUserInfo != null;
+                return autoconfigUserInfo;
+            } else {
+                return bootstrapUserInfo.deepClone();
+            }
         } else {
             return ReservedUserInfo.defaultEnabledUserInfo();
         }
```

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/esnative/NativeUsersStore.java
Comment (albertzaharovits) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/esnative/NativeUsersStore.java:303]: The get after a successful create is unnecessary.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/esnative/NativeUsersStore.java
Comment (albertzaharovits) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/esnative/NativeUsersStore.java:292]: This method overall is poorly designed.

There is a very similar private method above. This one here chains that one with another one, for no apparent reason, to me, at least. Afterwards it is going to be used in a very particular circumstance (creating the automatically configured elastic user), but its interface is designed to  be used in general, which goes against `private` qualifier of the method above.

Please follow my suggestion from https://github.com/elastic/elasticsearch/pull/77036/files#r701072274

**Review by albertzaharovits — COMMENTED** [review]
The logic looks OK, the placement of it definitely needs a more careful consideration (eg we shouldn't do a store in a get).

I've left a rough code suggestion in https://github.com/elastic/elasticsearch/pull/77036/files#r701072274

**Comment by BigPandaToo** [comment]
@elasticmachine update branch

**Comment by BigPandaToo** [comment]
@elasticmachine update branch

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/esnative/ReservedRealm.java
Comment (BigPandaToo) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/esnative/ReservedRealm.java:92]: I have refactored it

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/esnative/NativeUsersStore.java
Comment (BigPandaToo) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/esnative/NativeUsersStore.java]: See ^^

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/esnative/ReservedRealm.java
Comment (jkakavas) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/esnative/ReservedRealm.java]: I think the main problem with leaving this to the caller of `getUserInfo` is that it has no way of knowing if the `ReservedUserInfo` it got is from the keystore or the index

**Comment by BigPandaToo** [comment]
@elasticmachine update branch

**Comment by BigPandaToo** [comment]
@elasticmachine update branch

File: x-pack/plugin/security/src/internalClusterTest/java/org/elasticsearch/xpack/security/authc/esnative/PasswordHashPromotionIntegTests.java
Comment (BigPandaToo) [reply x-pack/plugin/security/src/internalClusterTest/java/org/elasticsearch/xpack/security/authc/esnative/PasswordHashPromotionIntegTests.java:70]: Just validating that now when the index exists we can authenticate (not sure it is necessary...)

**Comment by BigPandaToo** [comment]
@elasticmachine update branch

**Comment by BigPandaToo** [comment]
@elasticmachine update branch

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/esnative/ReservedRealm.java
Comment (jkakavas) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/esnative/ReservedRealm.java:232]: can we factor out 
```
if (autoconfigured && username.equals(ElasticUser.NAME) && bootstrapUserInfo.verifyPassword(credentials)) {
                        nativeUsersStore.storeAutoconfiguredElasticUser(bootstrapUserInfo, listener);
                    } else {
                        listener.onResponse(getDefaultUserInfo(username));
                    }
```

instead of copying it as I had suggested earlier on too ?

File: x-pack/plugin/security/src/internalClusterTest/java/org/elasticsearch/xpack/security/authc/esnative/PasswordHashPromotionIntegTests.java
Comment (jkakavas) [inline x-pack/plugin/security/src/internalClusterTest/java/org/elasticsearch/xpack/security/authc/esnative/PasswordHashPromotionIntegTests.java:41]: do we need this ?

**Review by jkakavas — COMMENTED** [review]
We need packaging tests for this for sure but I will add some in a follow up

File: x-pack/plugin/security/src/internalClusterTest/java/org/elasticsearch/xpack/security/authc/esnative/PasswordHashPromotionIntegTests.java
Comment (BigPandaToo) [reply x-pack/plugin/security/src/internalClusterTest/java/org/elasticsearch/xpack/security/authc/esnative/PasswordHashPromotionIntegTests.java:30]: Added one where the index exists, but the elasticuser doesn't. 
`securityIndexExistsAuthenticate` is where elasticuser already exists

File: x-pack/plugin/security/src/internalClusterTest/java/org/elasticsearch/xpack/security/authc/esnative/PasswordHashPromotionIntegTests.java
Comment (BigPandaToo) [reply x-pack/plugin/security/src/internalClusterTest/java/org/elasticsearch/xpack/security/authc/esnative/PasswordHashPromotionIntegTests.java:41]: Not really

**Comment by BigPandaToo** [comment]
@elasticmachine update branch

File: x-pack/plugin/security/src/internalClusterTest/java/org/elasticsearch/xpack/security/authc/esnative/PasswordHashAndBootstrapPwdInterTests.java
Comment (jkakavas) [inline x-pack/plugin/security/src/internalClusterTest/java/org/elasticsearch/xpack/security/authc/esnative/PasswordHashAndBootstrapPwdInterTests.java:29]: PasswordHashAndBootstrapPasswordIntegTests

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/esnative/ReservedRealm.java
Comment (jkakavas) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/esnative/ReservedRealm.java:240]: Naming is hard.. does this get the/a Auto Configured user ?

File: x-pack/plugin/security/src/internalClusterTest/java/org/elasticsearch/xpack/security/authc/esnative/PasswordHashPromotionIntegTests.java
Comment (jkakavas) [reply x-pack/plugin/security/src/internalClusterTest/java/org/elasticsearch/xpack/security/authc/esnative/PasswordHashPromotionIntegTests.java:37]: >  securityIndexExistsAuthenticate is where elasticuser already exists

But this wasn't a test and we removed it. I was looking more for a test where we index the document for the elastic user manually and then we try to authenticate with the autoconfigured password hash and it fails. If you think this is easy/quick enough to add you can do, otherwise leave a TODO and we'll follow up, let's not block this.

**Review by jkakavas — APPROVED** [review]
LGTM , I left some comments for consideration but won't block. Feel free to merge this once CI is green

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/esnative/NativeUsersStore.java
Comment (jkakavas) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/esnative/NativeUsersStore.java:263]: This needs to be INDEX, not UPDATE

**Review by jkakavas — CHANGES_REQUESTED** [review]
https://github.com/elastic/elasticsearch/pull/77036/files#r703651072 needs to be addressed before merging

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/esnative/NativeUsersStore.java
Comment (BigPandaToo) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/esnative/NativeUsersStore.java:263]: Just figured it out!

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/esnative/ReservedRealm.java
Comment (BigPandaToo) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/esnative/ReservedRealm.java:240]: You are right, not necessary autoconfigured.  May be `getAutoconfiguredOrDefaultUser`

**Comment by BigPandaToo** [comment]
@elasticmachine run elasticsearch-ci/part-2

**Comment by BigPandaToo** [comment]
@elasticmachine run elasticsearch-ci/part-2

**Comment by BigPandaToo** [comment]
@elasticmachine run elasticsearch-ci/part-2-fips

**Review by jkakavas — APPROVED** [review]
LGTM, lets hold off merging for now as we discussed

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/esnative/NativeUsersStore.java
Comment (jkakavas) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/esnative/NativeUsersStore.java:293]: Do we need this ? We only ever pass the onFailure of the ActionListener we also pass as a parameter, so we can just simply leave this as it was ?

Comment (jkakavas) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/esnative/NativeUsersStore.java:293]: Happy to push this change if needed, no worries

Comment (BigPandaToo) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/esnative/NativeUsersStore.java:293]: Please do! I merged the conflict, but there are some failures I don't have time to look at (my flight is in 2hr ))).

Comment (BigPandaToo) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/esnative/NativeUsersStore.java:293]: > Do we need this ? We only ever pass the onFailure of the ActionListener we also pass as a parameter, so we can just simply leave this as it was ?

I use it for the special error handling to return INTERNAL_SERVER_ERROR, look at `storeAutoconfiguredElasticUser`

**Comment by jkakavas** [comment]
@albertzaharovits we didn't have to merge this before you return so feel free to take another look when you're back and I can make all necessary changes and amendments while Lyudmila is away

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/esnative/NativeUsersStore.java
Comment (albertzaharovits) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/esnative/NativeUsersStore.java:328]: The logic here is wrong, I did some testing.
Firstly, the branching on the instance type should be on the `listener::onFailure` argument, not on the `consumer`. But even then, intercepting `VersionConflictEngineException` exceptions in order to return 401 instead of 500 is not correct. It is possible that multiple concurrent requests **on the same node** try to set the auto configured elastic user password symultaneously, and only one will succeed, but the others will be returned with 401 which is wrong (should signal authn success).

This is tricky I need to think some more for the proper fix.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/DefaultAuthenticationFailureHandler.java
Comment (albertzaharovits) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/DefaultAuthenticationFailureHandler.java:173]: Seeing this code, I'm having second thoughts about the plan here.
I didn't realize we return 401 as a catch all error case.
From this perspective, it makes little sense to return 401 if, eg, the index cannot be created, but return 500 if the elastic password had already been changed.

Comment (jkakavas) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/DefaultAuthenticationFailureHandler.java:173]: > From this perspective, it makes little sense to return 401 if, eg, the index cannot be created, but return 500 if the elastic password had already been changed.

It's the other way round, right ?

We can simplify this very much by just returning 401 in all of our cases too. What we were trying to capture was a difference between:

- "we checked, this is the wrong password - not the one in the keystore" (401) 
- "we checked, this is the wrong password - it's the one in the local keystore but something else has set it on the index" (401) 
- "we checked, this is the correct password but we can't persist it and some other node might set it to something else later, so next time you call it might be a 401 so it doesn't make sense to return 200" (500)

The contract to the user is that 
1) if they successfully authenticate once then they can successfully authenticate with these credentials until something (not ES internal) changes the password. 
2) If they get a 401 then the password is wrong and they should not try it again. 

but maybe we're thinking this too much from the user's perspective? There is no mandate that a 401 is permanent or that the user should not retry the request etc. We were trying to use this to signal our logic.

Comment (albertzaharovits) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/DefaultAuthenticationFailureHandler.java:173]: > We can simplify this very much by just returning 401 in all of our cases too. What we were trying to capture was a difference between:
> 1)  "we checked, this is the wrong password - not the one in the keystore" (401)
> 2)  "we checked, this is the wrong password - it's the one in the local keystore but something else has set it on the index" (401)
> 3)  "we checked, this is the correct password but we can't persist it and some other node might set it to something else later, so next time you call it might be a 401 so it doesn't make sense to return 200" (500)

Thanks for writing this down!
I think this captures the intent when we set about to code this.

On point 2) above. That something else that already set the password could be an identical process racing to set the password after a successful validation of the hash in the keystore. The racing process could be on the same node, writing the same password hash in the index, or on a different node, potentially writing a different one, or it can even be a password change by the user through the API or the cmd line tool. In this general case (without knowing who set the elastic user in the index in the meantime) returning 401 or 200 are both wrong (the only two error codes we can currently return for authn failures). Authentication has to be retried, given the new state where there is an elastic user in the .security index (which would make authn validate that instead of the keystore value).
We either retry it internally or we signal it to the client to retry it themself.

Point 3) sounds right to me, apart that maybe we should return 503. But I missed that there are a bunch of other error conditions that should probably be 503 and are now 401, eg shards unavailable for the .security index (though those are "read" ops and we're going to return 503 for "write" ones).

-------

Overall, after this conversation, I still think returning 503 if the "promised" password cannot be persisted (because the index or document cannot be created, **or** because they have been created in the meantime by something else) is the best course of action, because this "write" op during authentication is the only exception (which is not the approach this PR implements, FWIW).
I can also concede to returning 401 as well, given that the other "reads" that fail for system reasons also return 401.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/esnative/NativeUsersStore.java
Comment (albertzaharovits) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/esnative/NativeUsersStore.java:328]: From https://github.com/elastic/elasticsearch/pull/77036#discussion_r708841870 , I think we should return 503 (slightly more suggestive than 500, I think) when the promised password cannot be persisted. This is what we set about to do, though back then I haven't considered that any other failures return 401.
Also, in particular, returning 401 because the elastic user becomes present in the .security index in the meantime, is technically wrong because it assumes the update must be with a different password, which might be true in practice, but only because of reserved realm caching, which means that the update comes from a different node, or the user, which probably use a different password to set for elastic, but not necessarily.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/esnative/ReservedRealm.java
Comment (albertzaharovits) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/esnative/ReservedRealm.java:246]: This condition here makes it so that the auto configured elastic password is actually checked twice (second time in ReservedRealm#doAuthenticate) when wrong.
Not a big issue in practice, but it shows that the control flow is not polished.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/esnative/ReservedRealm.java
Comment (albertzaharovits) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/esnative/ReservedRealm.java:226]: I still don't like that a "get" does a "set", and that it takes a credentials parameter for it, which can be null but then the callees can't deal with a null parameter: this code will fail the lookup of the `elastic` user if it hasn't been created yet.

**Comment by albertzaharovits** [comment]
Closed in favor of https://github.com/elastic/elasticsearch/pull/78306 .

**Pull request closed without merging**
